src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -345,7 +347,8 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool whenIsKeyword, Precedence
                 TypeSyntax type = null;
                 if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
                       // If it is a nameof, skip the 'if' and parse as an expression. 
-                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken))
+                      (this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword || this.PeekToken(1).Kind != SyntaxKind.OpenParenToken) ||
+                    LooksLikeTupleArrayType())
```

Comment:
The parentheses around the subexpressions before `LooksLikeTupleArrayType()` are slightly different from `ParseTypeOrPatternForIsOperatorCore`. Is that intentional? If not, should this entire `if` condition be extracted to a helper method?